tests/pytests/functional/transport/rabbitmq/test_pub_server_channel.py

```diff
@@ -0,0 +1,221 @@
+import ctypes
+import logging
+import multiprocessing
+import signal
+import time
+
+import pytest
+import salt.config
+import salt.exceptions
+import salt.ext.tornado.gen
+import salt.ext.tornado.ioloop
+import salt.log.setup
+import salt.transport.client
+import salt.transport.server
+import salt.utils.platform
+import salt.utils.process
+import salt.utils.stringutils
+from saltfactories.utils.processes import terminate_process
+
+log = logging.getLogger(__name__)
+
+
+class Collector(salt.utils.process.SignalHandlingProcess):
+    """
+    A process for collecting published events
+    """
+    def __init__(
+        self, minion_config, pub_uri, aes_key, timeout=30
+    ):
+        super().__init__()
+        self.minion_config = minion_config
+        self.pub_uri = pub_uri
+        self.aes_key = aes_key
+        self.timeout = timeout
+        self.hard_timeout = time.time() + timeout + 30
+        self.manager = multiprocessing.Manager()
+        self.results = self.manager.list()
+        self.stopped = multiprocessing.Event()
+        self.started = multiprocessing.Event()
+        self.running = multiprocessing.Event()
+
+    def run(self):
+        # receive
+        rmq_connection_wrapper = salt.transport.rabbitmq.RMQBlockingConnectionWrapper(self.minion_config)
+
+        """
+        Gather results until then number of seconds specified by timeout passes
+        without receiving a message
+        """
+        last_msg = time.time()
+        serial = salt.payload.Serial(self.minion_config)
+        crypticle = salt.crypt.Crypticle(self.minion_config, self.aes_key)
+        self.started.set()
+        while True:
+            curr_time = time.time()
+            if time.time() > self.hard_timeout:
+                break
+            if curr_time - last_msg >= self.timeout:
+                break
```

Comment:
**question** I've seen this code in the `test_async_pub_channel` - would it be possible to refactor this somewhat?

Also, I wonder if it's possible to extract some/most/all of this functionality into the `salt.transport.rabbitmq` layer. If there's a lot of commonality between spinning up a pub/sub, especially if the only real difference is how the messages are handled.

Reply:
These tests were adopted from similar tests that test zeromq transport. 
There's plenty of opportunity to refactor and eliminate some code duplication. Note: one school of thought on the Java side of things is that duplication in test code is more acceptable than product code. However, I'll try to refactor this more.